enhancements/microshift/workload-partitioning-in-microshift.md

```diff
@@ -0,0 +1,313 @@
+
+- kubelet configuration to enable and configure CPU Manager for the workloads.
+
+  
+This will limit the cores allowed to run microshift control plane services, maximizing the CPU core for application payloads.
+
+### Workflow Description
+
+#### System and MicroShift configuration
+
+##### OSTree
+1. User supplies Microshift configuration  using blueprint
+    -  /etc/microshift/config.yaml - embed Kubelet configuration through microshift config.
+    - /etc/kubernetes/openshift-workload-pinning - kubelet configuration file.
+1. User supplies CRIO configuration file using blueprint.
+1. User specify CPUAffinity with systemd drop-in  configuration using blueprint,
+   for those services:
+    -  crio
+    -  Microshift 
+
+
+1. User builds the blueprint
+1. User deploys the commit / installs the system.
+1. System boots
+
+Example blueprint:
+```
+name = "microshift-workload-partiontining"
+version = "0.0.1"
+modules = []
+groups = []
+distro = "rhel-94"
+
+[[packages]]
+name = "microshift"
+version = "4.17.*"
+
+[[customizations.services]]
+enabled = ["microshift"]
+
+[[customizations.files]]
+path = "/etc/microshift/config.yaml"
+data = """
+kubelet:
+  reservedSystemCPUs: "0,6,7"
+  cpuManagerPolicy: static
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true"
+  cpuManagerReconcilePeriod: 5s
+"""
+
+[[customizations.files]]
+path = "/etc/crio/crio.conf.d/20-microshift-wp.conf"
+data = """
+[crio.runtime]
+infra_ctr_cpuset = "0,6,7"
+
+[crio.runtime.workloads.management]
+activation_annotation = "target.workload.openshift.io/management"
+annotation_prefix = "resources.workload.openshift.io"
+resources = { "cpushares" = 0, "cpuset" = "0,6,7" }
+"""
+
+[[customizations.files]]
+path = "/etc/systemd/system/crio.service.d/microshift-cpuaffinity.conf"
+data = """
+[Service]
+CPUAffinity=0,6,7
+"""
+
+[[customizations.files]]
+path = "/etc/systemd/system/microshift.service.d/microshift-cpuaffinity.conf"
+data = """
+[Service]
+CPUAffinity=0,6,7
+"""
+
+
+[[customizations.files]]
+path = "/etc/kubernetes/openshift-workload-pinning"
+data = """
+{
+  "management": {
+    "cpuset": "0,6,7"
+  }
+}
+"""
+
+```
+
+##### RPM
+1. User creates configuration files with the workload partitioning configuration
+1. reboot the host
+
+### API Extensions
+
+Following API extensions are expected:
+- A passthrough from MicroShift's config to Kubelet config.
+
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+Purely MicroShift enhancement.
+
+### Implementation Details/Notes/Constraints
+
+#### Kubelet configuration 
+- instructs kubelet to modify the node resource with the capacity and allocatable CPUs.
+  
+  add to the file /etc/kubernetes/openshift-workload-pinning
+  ```json
+  {
+    "management": {
+      "cpuset": "0,6,7"
```

Comment:
It would be especially friendly if the user could specify the reserved CPUs in one file and it propagates to all the necessary configs.

Reply:
I agree that this sounds desirable. @eslutsky - can you comment please?

Reply:
as a  principle  we Dont  change OS  configuration from Microshift runtime, so  all  systemd/OS related configuration values has to be configured externally.